tune_llama.py

```diff
@@ -0,0 +1,285 @@
+# Finetunes LLaMA-7B on Alpaca dataset. Supports DDP via torchrun.
+# Disclaimer: Referenced https://github.com/tatsu-lab/stanford_alpaca/tree/main/train.py
+# Ex: torchrun --nproc_per_node 4 master_port 25555 tune_llama.py --dp_size 2 --tp_size 2
+import os
+from dataclasses import dataclass, field
+from typing import Dict, List, Optional
+
+import torch
+import torch.distributed as dist
+import torch.nn.functional as F
+import tqdm
+import transformers
+from torch.nn.parallel import DistributedDataParallel as DDP
+from torch.utils.data import DataLoader
+from torch.utils.data.distributed import DistributedSampler
+from transformers import (LlamaConfig, LlamaTokenizer, PreTrainedTokenizer,
+                          TrainingArguments, get_cosine_schedule_with_warmup)
+
+from src.data_utils import *
+from src.modeling_llama import LlamaForCausalLM
+
+
+def manual_reduction(model, group: dist.ProcessGroupNCCL, average: bool = False):
+    """Manually reduce the gradient within a group. I suspect DDP's backward is buggy when there are multiple process groups."""
+    for param in model.parameters():
+        if param.requires_grad and param.grad is not None:
+            # All-reduce gradients manually within the specified group
+            dist.all_reduce(param.grad.data, op=dist.ReduceOp.SUM, group=group)
+            if average:
+                param.grad.data /= dist.get_world_size(group)
+
+
+def init_dist(dp_size: int = -1) -> List[dist.ProcessGroupNCCL]:
+    dp_size = int(os.environ["WORLD_SIZE"]) if dp_size == -1 else dp_size
+    rank = int(os.environ["RANK"])
+    local_rank = int(os.environ["LOCAL_RANK"])
+    world_size = int(os.environ["WORLD_SIZE"])
+
+    dist.init_process_group(world_size=world_size, rank=rank, init_method="env://", backend="nccl")
+    torch.cuda.set_device(local_rank)
+
+    # Set up Tensor Parallel within each Data Parallel group
+    tp_size = world_size // dp_size
+    if dp_size > 1:
+        tp_groups = [
+            dist.new_group(ranks=range(group_id * tp_size, (group_id + 1) * tp_size)) for group_id in range(dp_size)
+        ]
+        return tp_groups
```

Comment:
You should create both tp group and dp group. Only return the group which current process belongs to.

Reply:
Cool